capella_model_explorer/backend/model_diff.py

```diff
@@ -0,0 +1,485 @@
+    """The revision hash."""
+    revision: str
+    """The original revision passed to the diff tool."""
+    author: str
+    """The author of the revision, in "Name <email@domain>" format."""
+    date: datetime.datetime
+    """The time and date of the revision."""
+    description: str
+    """The description of the revision, i.e. the commit message."""
+    tag: str | None
+    """The tag of the commit."""
+
+
+class Metadata(te.TypedDict):
+    model: dict[str, t.Any]
+    """The 'modelinfo' used to load the models, sans the revision key."""
+    new_revision: RevisionInfo
+    old_revision: RevisionInfo
+
+
+class BaseObject(te.TypedDict):
+    uuid: str
+    display_name: str
+    """Name for displaying in the frontend.
+
+    This is usually the ``name`` attribute of the "current" version of
+    the object.
+    """
+
+
+class FullObject(BaseObject, te.TypedDict):
+    attributes: dict[str, t.Any]
+    """All attributes that the object has (or had)."""
+
+
+class ChangedAttribute(te.TypedDict):
+    previous: t.Any
+    """The old value of the attribute."""
+    current: t.Any
+    """The new value of the attribute."""
+
+
+class ChangedObject(BaseObject, te.TypedDict):
+    attributes: dict[str, ChangedAttribute]
+    """The attributes that were changed."""
+
+
+class ObjectChange(te.TypedDict, total=False):
+    created: list[FullObject]
+    """Contains objects that were created."""
+    deleted: list[FullObject]
+    """Contains objects that were deleted."""
+    modified: list[ChangedObject]
+
+
+ObjectLayer: te.TypeAlias = "dict[str, ObjectChange]"
+
+
+class ObjectChanges(te.TypedDict, total=False):
+    oa: ObjectLayer
+    """Changes to objects from the OperationalAnalysis layer."""
+    sa: ObjectLayer
+    """Changes to objects from the SystemAnalysis layer."""
+    la: ObjectLayer
+    """Changes to objects from the LogicalAnalysis layer."""
+    pa: ObjectLayer
+    """Changes to objects from the PhysicalAnalysis layer."""
+    epbs: ObjectLayer
+    """Changes to objects from the EPBS layer."""
+
+
+class ChangeSummaryDocument(te.TypedDict):
+    metadata: Metadata
+    objects: ObjectChanges
+
+
+def init_model(model: capellambse.MelodyModel):
+    file_handler = model.resources["\x00"]
+    path = str(file_handler.path)
+    model_data: dict = {
+        "metadata": {"model": {"path": path, "entrypoint": None}},
+        "diagrams": {
+            "created": {},
+            "modified": {},
+            "deleted": {},
+        },
+        "objects": {
+            "created": {},
+            "modified": {},
+            "deleted": {},
+        },
+    }
+
+    if isinstance(file_handler, git.GitFileHandler):
+        path = str(file_handler.cache_dir)
+    elif (
+        isinstance(file_handler, local.LocalFileHandler)
+        and file_handler.rootdir.joinpath(".git").is_dir()
+    ):
+        pass
+    else:
+        return {"error": "Not a git repo"}, model_data
+    return path, model_data
```

Comment:
The error case here isn't handled properly, in neither of the two callers to `init_model`.

- `populate_commits` checks `if "error" in result`, which will falsely trigger if the returned path contains the word "error" anywhere.
- `get_diff_data` doesn't handle the error case at all.

There's really no need for returning a dict with a reason if there can only be one reason anyways. We can save a lot of headache by either:

- returning `None` in case of errors (then it's an unambiguous `if path is not None` check at the call site), or
- raising an exception on error with more information, which can then be caught further up the stack (e.g. by responding with an API error)